### PR TITLE
[CoAP] Adding CLI commands

### PR DESCRIPTION
**Pull request by Gr3yh0und**
Adds CLI commands for a simple CoAP client and server with one resource called "test". Supports Non- and Confirmable requests...

I was playing around and thought due to the many requests on the google group it might be useful to have this functionality as an example working "out of the box". It's by far not finished but I just wanted to see if something like is wanted/useful.  And I'm also just getting warm with C again...

### CODE REVIEW

File: src/cli/cli.cpp [hunk 1]
```diff
@@ -602,6 +612,259 @@ void Interpreter::ProcessChildTimeout(int argc, char *argv[])
...
+    // CoAP-Type
+    if (strcmp(argv[3], "Confirmable") == 0)
+    {
+        coapType = kCoapTypeConfirmable;
+    }
+
+    // Content
+    if (argv[4])
+    {
+        strcpy(coapContent, argv[4]);
```
Comment (LuDuda) [inline src/cli/cli.cpp]: If coapContent has fixed size then you need a check if` argv[4] `does not exceed it.

But anyhow i think you can remove temporary `coapContent` buffer and use `otMessageAppend ` that you have in line 816 to append string directly to the message buffer.

File: src/cli/cli.cpp [hunk 2]
```diff
@@ -602,6 +612,259 @@ void Interpreter::ProcessChildTimeout(int argc, char *argv[])
...
+    }
+    else if (strcmp(argv[0], "stop") == 0)
+    {
+        otCoapServerRemoveResource(mInstance, &mResource);
+        SuccessOrExit(error = otCoapServerStop(mInstance));
+        sServer->OutputFormat("Stopping server ... ");
+    }
+    else if (strcmp(argv[0], "name") == 0)
+    {
+        if (argv[1]){
```
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: I found that in the contribution readme as well but I couldn't find out on which target/dir I needed to run that. Can you please give me a heads up? Thanks.

Comment (LuDuda) [inline src/cli/cli.cpp]: Sure no problem!

```
$ ./bootstrap
$ ./configure --enable-ftd --enable-cli --enable-diag --enable-dhcp6-client --enable-dhcp6-server --enable-dns-client --enable-commissioner --enable-joiner --with-examples=posix
$ make pretty
```

If you are interest you can check how travis does all the checks in this file:  [script.sh](https://github.com/openthread/openthread/blob/master/.travis/script.sh)

**Comment by Gr3yh0und** [comment]
@nibanks : Yeah, shouldn't be a problem. Do you or @jwhui have a suggesstion for the directory/file structure?

File: src/cli/cli.cpp (hunk 2 above)
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: Thanks for that! Haven't tried the posix platform earlier on...

File: src/cli/cli.cpp [hunk 3]
```diff
@@ -602,6 +612,259 @@ void Interpreter::ProcessChildTimeout(int argc, char *argv[])
...
+            mResource.mUriPath = argv[1];
```
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: Due to the definition mResource is persistent and therefore mUriPath as well. At least it works as supposed?

File: src/cli/cli.cpp [hunk 4]
```diff
@@ -602,6 +612,259 @@ void Interpreter::ProcessChildTimeout(int argc, char *argv[])
...
+    otCoapHeader header;
+    const otNetifAddress *address;
+
+    // Default parameters
+    char coapUri[] = "test";
+    char coapContent[20] = "";
+    otCoapType coapType = kCoapTypeNonConfirmable;
+    otCoapCode coapCode = kCoapRequestGet;
+    otIp6Address coapDestinationIp;
+    otIp6AddressFromString("FF01::1", &coapDestinationIp); // Broadcast
```
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: Removed the Broadcast as it is truely not very useful in that use case.

File: src/cli/cli.cpp [hunk 5]
```diff
@@ -602,6 +612,259 @@ void Interpreter::ProcessChildTimeout(int argc, char *argv[])
...
+        default:
+            sServer->OutputFormat("Undefined\r\n");
+            return;
+    }
+
+    // Output payload of message if existent
+    uint16_t payloadLength = otMessageGetLength(aMessage) - otMessageGetOffset(aMessage);
+    if (payloadLength > 0)
+    {
+        char payload[payloadLength] = {};
```
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: Could you explain that a little further please?

File: src/cli/cli.cpp (hunk 3 above)
Comment (LuDuda) [inline src/cli/cli.cpp]: What i ment is that mUriPath is a pointer to constant char `const char *mUriPath`. The value of argv[1] is temporary allocated on the memory that will change over time. Thus if you want to have it configurable you need to reserve dedicated memory for CoAP Resource URI.

File: src/cli/cli.cpp (hunk 5 above)
Comment (LuDuda) [inline src/cli/cli.cpp]: Sure, i was thinking about something like this:
https://github.com/openthread/openthread/blob/master/src/core/common/message.cpp#L670
https://github.com/openthread/openthread/blob/master/src/cli/cli.cpp#L2980

File: src/cli/cli.cpp (hunk 3 above)
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: Ah, got it now. Will fix that in the next commit.

File: src/cli/cli.cpp (hunk 5 above)
Comment (Gr3yh0und) [inline src/cli/cli.cpp]: Thanks, I'll have a look at it.

**Comment by jwhui** [comment]
I think it makes sense to separate this out into a separate file.  I think we need to generally clean the CLI component up.  But, for now, you could reference [cli_dataset.cpp](https://github.com/openthread/openthread/blob/master/src/cli/cli_dataset.cpp) as an example.  Basically, have a top-level CLI in `cli.cpp` that calls into something like `cli_coap.cpp`.

**Comment by turon** [comment]
Another paradigm that can work with the current structure with minimal changes is to keep the ifdefs in cli.hpp to define `ProcessCoapServer` and `ProcessCoapClient`, keep the ifdefs in cli.cpp to map the command strings to those calls, and then simply move your current implementation of `Cli::ProcessCoapServer` and `Cli::ProcessCoapServer` into a new file named cli_coap.cpp.  Then in Makefile.am, only include cli_coap.cpp based on the ifdef.

**Comment by jwhui** [comment]
@turon, yes, that's exactly how the example in my prior comment works.

**Comment by turon** [comment]
@jwhui, sure, just saying you don't need to restructure everything to implement a second sub-Interpreter class as `cli_dataset.cpp` does, and you may not need a `cli_coap.hpp`.

**Comment by jwhui** [comment]
@turon, benefit of creating a new class is that you can encapsulate related member variables and reduce the number of overall `#ifdef`.

**Comment by Gr3yh0und** [comment]
I've moved everything into a new CLI class in new files and fixed most of the comments from @LuDuda . Looking forward to further comments!

File: src/cli/cli_coap.cpp [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+        }
+    }
+    else
+    {
+        ExitNow(error = kThreadError_InvalidArgs);
+    }
+
+exit:
+    (void)argc;
+    (void)argv;
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: These should not be needed - the variables are used above.

File: src/cli/cli_coap.cpp [hunk 7]
```diff
@@ -0,0 +1,356 @@
...
+    {
+        if ((otCoapHeaderGetType(aHeader) == kCoapTypeConfirmable) || otCoapHeaderGetCode(aHeader) == kCoapRequestGet)
+        {
+            sServer->OutputFormat("CoAP response sent successfully!\r\n");
+        }
+    }
+
+    (void)aHeader;
+    (void)aMessage;
+    (void)aMessageInfo;
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: These should not be needed.  Variables are used above.

File: src/cli/cli_coap.cpp [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+
+ThreadError Coap::ProcessClient(int argc, char *argv[])
+{
+    ThreadError error = kThreadError_None;
+    otMessage *message;
+    otMessageInfo messageInfo;
+    otCoapHeader header;
+
+    // Default parameters
+    char coapUri[] = "test";
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: Declare as `static const`.

File: src/cli/cli_coap.cpp [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+    messageInfo.mPeerAddr = coapDestinationIp;
+    messageInfo.mPeerPort = OT_DEFAULT_COAP_PORT;
+    messageInfo.mInterfaceId = OT_NETIF_INTERFACE_ID_THREAD;
+
+    error = otCoapSendRequest(sInstance, message, &messageInfo,
+                              (otCoapResponseHandler) &Coap::s_HandleClientResponse, NULL);
+
+    sServer->OutputFormat("Sending CoAP message: ");
+
+    if (error != kThreadError_None)
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: Move below `exit:` label.

File: src/cli/cli_coap.cpp [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+    uint8_t bytesPrinted = 0;
+
+    if (payloadLength > 0)
+    {
+        char payload[outputBufferLength] = {};
+        while (bytesPrinted < payloadLength)
+        {
+            otMessageRead(aMessage, otMessageGetOffset(aMessage) + bytesPrinted, &payload, outputBufferLength);
+            bytesPrinted = bytesPrinted + outputBufferLength;
+            sServer->OutputFormat(" payload: %s", payload);
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: This assumes `payload` is NULL-terminated.  Do we want to print out the raw bytes instead?

File: src/cli/cli_coap.cpp [hunk 11]
```diff
@@ -0,0 +1,356 @@
...
+    uint8_t payloadLength = otMessageGetLength(aMessage) - otMessageGetOffset(aMessage);
+    uint8_t outputBufferLength = 16;
+    uint8_t bytesPrinted = 0;
+
+    if (payloadLength > 0)
+    {
+        char payload[outputBufferLength] = {};
+        while (bytesPrinted < payloadLength)
+        {
+            otMessageRead(aMessage, otMessageGetOffset(aMessage) + bytesPrinted, &payload, outputBufferLength);
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: What if the message is not a multiple of `outputBufferLength`?

File: src/cli/cli_coap.cpp [hunk 12]
```diff
@@ -0,0 +1,356 @@
...
+    }
+
+    // CoAP-Type
+    if ((argc > 3) && (strcmp(argv[3], "Confirmable") == 0))
+    {
+        coapType = kCoapTypeConfirmable;
+    }
+
+    otCoapHeaderInit(&header, coapType, coapCode);
+    otCoapHeaderGenerateToken(&header, 2);
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: Avoid magic numbers.

File: src/cli/cli_coap.cpp [hunk 13]
```diff
@@ -0,0 +1,356 @@
...
+exit:
+
+    if (error != kThreadError_None && responseMessage != NULL)
+    {
+        sServer->OutputFormat("Cannot send CoAP response message: Error %d\r\n", error);
+        otMessageFree(responseMessage);
+    }
+    else
+    {
+        if ((otCoapHeaderGetType(aHeader) == kCoapTypeConfirmable) || otCoapHeaderGetCode(aHeader) == kCoapRequestGet)
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: How about simplifying to `otCoapHeaderGetType(aHeader) >= kCoapResponseCodeMin`?

**Comment by jwhui** [comment]
Few more comments:

1. Update the Makefile to include the new source files.
2. Update the Travis scripts to include this in the `pretty-check` target and build.
3. Update the [README.md](https://github.com/openthread/openthread/blob/master/src/cli/README.md) to include the new commands and example output.

File: src/cli/cli_coap.cpp [hunk 14]
```diff
@@ -0,0 +1,356 @@
...
+    {
+        otCoapServerRemoveResource(sInstance, sResource);
+        SuccessOrExit(error = otCoapServerStop(sInstance));
+        sServer->OutputFormat("Server stopped: ");
+    }
+    else if (strcmp(argv[0], "name") == 0)
+    {
+        if (argc > 1)
+        {
+            sResource->mUriPath = argv[1];
```
Comment (LuDuda) [inline src/cli/cli_coap.cpp]: This still will not work. You need to declare some buffer for Uri Path and assign it to mUriPath. Then if you want to change name - change content of the buffer.

File: src/cli/cli_coap.cpp (hunk 8 above)
Comment (Gr3yh0und) [inline src/cli/cli_coap.cpp]: But this is going to be changed in line 284?

Comment (jwhui) [inline src/cli/cli_coap.cpp]: Looking at this again, you'll want to do a couple things:
1. Specify the maximum input length to support.
2. Use the safe version of `strcpy()`, namely `strlcpy()`.

File: src/cli/cli_coap.cpp (hunk 11 above)
Comment (Gr3yh0und) [inline src/cli/cli_coap.cpp]: I'm not sure what or how to do something here. As far as I checked the code some kind of buffer values are always used, like the example https://github.com/openthread/openthread/blob/master/src/cli/cli.cpp#L2980 @LuDuda pointed out.

File: src/cli/cli_coap.cpp (hunk 10 above)
Comment (Gr3yh0und) [inline src/cli/cli_coap.cpp]: Depends what you prefer. I copied the OutputBytes function from https://github.com/openthread/openthread/blob/master/src/cli/cli.cpp#L2980 first, but I liked seeing the "text" rather than HEX.

File: src/cli/cli_coap.cpp (hunk 9 above)
Comment (Gr3yh0und) [inline src/cli/cli_coap.cpp]: I have some trouble here as `message` is not initialized. So moving the `if` below the exit could lead to execution without `message` being initialized...

**Comment by Gr3yh0und** [comment]
I've made some changes and added README and Makefile as well @jwhui. I'm not sure on the travis side but as the whole `cli` folder is party of the `PRETTY_DIR` this should be fine now.

File: src/cli/cli_coap.cpp (hunk 9 above)
Comment (jwhui) [inline src/cli/cli_coap.cpp]: You can initialize `message` to `NULL` upon declaration, then check if `message != NULL` before calling `otMessageFree()`.

File: src/cli/cli_coap.cpp [hunk 15]
```diff
@@ -0,0 +1,356 @@
...
+
+void Coap::PrintPayload(otMessage *aMessage)
+{
+    uint8_t payloadLength = otMessageGetLength(aMessage) - otMessageGetOffset(aMessage);
+    uint8_t outputBufferLength = kMaxBufferSize;
+    uint8_t bytesPrinted = 0;
+
+    if (payloadLength > 0)
+    {
+        char payload[outputBufferLength] = {};
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: Initializing a variable-length buffer like this won't work.  You should just declare `char payload[kMaxBufferSize]`.

File: src/cli/cli_coap.cpp (hunk 10 above)
Comment (jwhui) [inline src/cli/cli_coap.cpp]: I guess it really depends whether the CoAP payload is human-readable or binary.  Maybe we could add an option that allows the user to choose?

If we are printing a string (e.g. `%s`), we need to make sure the buffer is NULL-terminated.

File: src/cli/cli_coap.cpp (hunk 11 above)
Comment (jwhui) [inline src/cli/cli_coap.cpp]: This matters more if we are printing raw bytes.  See my response below.

File: src/cli/Makefile.am [hunk 16]
```diff
@@ -45,6 +45,10 @@ libopenthread_cli_a_SOURCES         = \
     cli_uart.cpp                      \
     cli_udp.cpp                       \
     $(NULL)
+    
+if OPENTHREAD_ENABLE_APPLICATION_COAP
+libopenthread_cli_a_SOURCES	+= cli_coap.cpp
```
Comment (jwhui) [inline src/cli/Makefile.am]: nit: align `=` with above and below.

File: src/cli/Makefile.am [hunk 17]
```diff
@@ -54,6 +58,10 @@ noinst_HEADERS                      = \
     cli_server.hpp                    \
     cli_udp.hpp                       \
     $(NULL)
+    
+if OPENTHREAD_ENABLE_APPLICATION_COAP
+noinst_HEADERS	+= cli_coap.hpp
```
Comment (jwhui) [inline src/cli/Makefile.am]: nit: align `=` with above and below.

File: src/cli/cli.cpp [hunk 18]
```diff
@@ -216,7 +222,7 @@ Interpreter::Interpreter(otInstance *aInstance):
 
 #if OPENTHREAD_ENABLE_DHCP6_CLIENT
     memset(mDhcpAddresses, 0, sizeof(mDhcpAddresses));
-#endif  // OPENTHREAD_ENABLE_DHCP6_CLIENT
+#endif // OPENTHREAD_ENABLE_DHCP6_CLIENT
```
Comment (jwhui) [inline src/cli/cli.cpp]: nit: unnecessary change

File: src/cli/cli_coap.cpp [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+        otCoapHeaderSetMessageId(&responseHeader, otCoapHeaderGetMessageId(aHeader));
+        otCoapHeaderSetToken(&responseHeader, otCoapHeaderGetToken(aHeader), otCoapHeaderGetTokenLength(aHeader));
+
+        if (otCoapHeaderGetCode(aHeader) == kCoapRequestGet)
+        {
+            otCoapHeaderSetPayloadMarker(&responseHeader);
+        }
+
+        responseMessage = otCoapNewMessage(sInstance, &responseHeader);
+        VerifyOrExit(responseMessage != NULL, error = kThreadError_Failed);
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: nit: `kThreadError_Failed` -> `kThreadError_NoBufs`

File: src/cli/cli_coap.cpp [hunk 20]
```diff
@@ -0,0 +1,356 @@
...
+        coapType = kCoapTypeConfirmable;
+    }
+
+    otCoapHeaderInit(&header, coapType, coapCode);
+    otCoapHeaderGenerateToken(&header, Thread::Coap::Header::kDefaultTokenLength);
+    SuccessOrExit(error = otCoapHeaderAppendUriPathOptions(&header, coapUri));
+    otCoapHeaderSetPayloadMarker(&header);
+
+    message = otCoapNewMessage(sInstance, &header);
+    VerifyOrExit(message != NULL, error = kThreadError_Failed);
```
Comment (jwhui) [inline src/cli/cli_coap.cpp]: nit: `kThreadError_Failed` -> `kThreadError_NoBufs`

File: src/cli/cli_coap.cpp [hunk 21]
```diff
@@ -0,0 +1,356 @@
...
+{
+    ThreadError error = kThreadError_None;
+
+    sInstance = aInstance;
+    sServer = &aServer;
+    sResource.mUriPath = sUriPath;
+    sResource.mContext = aInstance;
+    sResource.mHandler = (otCoapRequestHandler) &Coap::s_HandleServerResponse;
+
+    if (argc == 0)
```
Comment (Gr3yh0und) [inline src/cli/cli_coap.cpp]: Thanks, yeah that makes more sense.

**Comment by Gr3yh0und** [comment]
I've changed the output to HEX now (for the time being) and hopefully got all the other comments from @jwhui and @aeliot as well. I also made some small changes that I found annoying while testing...

File: src/cli/README.md [hunk 22]
```diff
@@ -269,6 +270,50 @@ Set the Thread Child Timeout value.
...
+* method: CoAP method to be used (GET/PUT/POST/DELETE) (mandatory).
+* IPv6address: IP address of the CoAP server to query (mandatory).
+* URI: URI String of the resource on the CoAP server (mandatory).
+* payload: In case of PUT/POST/DELETE a payload can be encapsulated (optional).
+* messageType: Switch between "Confirmable" and "Non-Confirmable"(default) (optional).
+
+```bash
+> coap client GET fe80::c8bc:678e:c80c:705b test
+Sending CoAP message: Done
+Received CoAP response with payload: 1234
```
Comment (jwhui) [inline src/cli/README.md]: Is this example consistent with the recent change to hex output?

File: src/cli/cli_coap.hpp [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+ * This class implements the CLI CoAP server and client.
+ *
+ */
+class Coap
+{
+public:
+    /**
+     * This method interprets a list of CLI arguments.
+     *
+     * @param[in]  argc  The number of elements in argv.
```
Comment (jwhui) [inline src/cli/cli_coap.hpp]: Add docs for `aInstance` and `aServer`.

File: src/cli/README.md (hunk 22 above)
Comment (Gr3yh0und) [inline src/cli/README.md]: You're right, I've added a new example.

**Comment by Gr3yh0und** [comment]
Thanks @jwhui for the input, I've updated the documentation.

File: src/cli/cli_coap.cpp [hunk 24]
```diff
@@ -0,0 +1,383 @@
...
+    {
+        strlcpy(coapUri, argv[2], kMaxUriLength);
+    }
+    else
+    {
+        ExitNow(error = kThreadError_InvalidArgs);
+    }
+
+    // CoAP-Type
+    if ((argc > 3) && (strcmp(argv[3], "Confirmable") == 0))
```
Comment (LuDuda) [inline src/cli/cli_coap.cpp]: nit: What about "con" instead of Confirmable?

File: src/cli/README.md [hunk 25]
```diff
@@ -269,6 +270,55 @@ Set the Thread Child Timeout value.
...
+```
+
+### coap client \<method\> \<IPv6address\> \<URI\> \<payload\> \<messageType\>
+
+Simple CoAP client that can send Non-/Confirmable GET/PUT/POST/DELETE messages.
+
+* method: CoAP method to be used (GET/PUT/POST/DELETE) (mandatory).
+* IPv6address: IP address of the CoAP server to query (mandatory).
+* URI: URI String of the resource on the CoAP server (mandatory).
+* payload: In case of PUT/POST/DELETE a payload can be encapsulated (optional).
```
Comment (LuDuda) [inline src/cli/README.md]: I might be wrong but i believe you can use this brackets: "[" and "]" to indicate that parameter is optional e.g. `[payload]` instead of `<payload>`

File: src/cli/README.md [hunk 26]
```diff
@@ -14,6 +14,7 @@ OpenThread test scripts use the CLI to execute test cases.
 * [child](#child-list)
 * [childmax](#childmax)
 * [childtimeout](#childtimeout)
+* [coap] (#coap)
```
Comment (LuDuda) [inline src/cli/README.md]: nit: Remove extra space that prevents to parse link correctly.

File: src/cli/cli_coap.cpp [hunk 27]
```diff
@@ -0,0 +1,383 @@
...
+            bytesPrinted += bytesToPrint;
+        }
+    }
+    sServer->OutputFormat("\r\n");
+}
+
+void Coap::ConvertToLower(char *aString)
+{
+    uint8_t i = 0;
+    while(aString[i])
```
Comment (LuDuda) [inline src/cli/cli_coap.cpp]: nit: Add extra space after while

**Comment by Gr3yh0und** [comment]
Thanks for the input @LuDuda, pushed a new commit.
